src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java

```diff
@@ -77,31 +81,27 @@ protected void runPipeline(final JavaSparkContext ctx) {
                 : IntervalUtils.getAllIntervalsForReference(readsHeader.getSequenceDictionary());
         JavaRDD<GATKRead> initialReads = readSource.getParallelReads(bam, intervals);
 
-        JavaRDD<GATKRead> markedReads = initialReads.map(new MarkDuplicatesStub());
+        JavaRDD<GATKRead> markedReads = MarkDuplicatesSpark.mark(initialReads, readsHeader, new OpticalDuplicateFinder(), 0);
         VariantsSparkSource variantsSparkSource = new VariantsSparkSource(ctx);
-        JavaRDD<Variant> bqsrKnownVariants = variantsSparkSource.getParallelVariants(baseRecalibrationKnownVariants);
 
-        GCSOptions options = PipelineOptionsFactory.as(GCSOptions.class);
-        options.setApiKey(apiKey);
+        // TODO: workaround for known bug in List version of getParallelVariants
+        if ( baseRecalibrationKnownVariants.size() > 1 ) {
+            throw new GATKException("Cannot currently handle more than one known sites file, " +
+                                    "as getParallelVariants(List) is broken");
+        }
+        JavaRDD<Variant> bqsrKnownVariants = variantsSparkSource.getParallelVariants(baseRecalibrationKnownVariants.get(0));
+
+        GCSOptions options = BucketUtils.getAuthenticatedGCSOptions(apiKey);
 
         final ReferenceDataflowSource referenceDataflowSource = new ReferenceDataflowSource(options, referenceURL, BaseRecalibratorDataflow.BQSR_REFERENCE_WINDOW_FUNCTION);
 
         // TODO: Look into broadcasting the reference to all of the workers. This would make AddContextDataToReadSpark
         // TODO: and ApplyBQSRStub simpler (#855).
         JavaPairRDD<GATKRead, ReadContextData> rddReadContext = AddContextDataToReadSpark.add(markedReads, referenceDataflowSource, bqsrKnownVariants);
-
-        JavaRDD<RecalibrationTables> tables = rddReadContext.map(new BaseRecalibratorStub());
-        RecalibrationTables nestedIntegerArrays = null;
-        if (tables != null) {
-            nestedIntegerArrays = tables.collect().get(0);
-        }
-        final Broadcast<RecalibrationTables> broadcast = ctx.broadcast(nestedIntegerArrays);
-
-        // ApplyBQSRStub.
-        JavaRDD<GATKRead> finalReads = markedReads.map(v1 -> {
-            RecalibrationTables value = broadcast.getValue();
-            return v1;
-        });
+        // TODO: broadcast the reads header?
+        final RecalibrationReport bqsrReport = BaseRecalibratorSparkFn.apply(rddReadContext, readsHeader, referenceDataflowSource.getReferenceSequenceDictionary(null), new RecalibrationArgumentCollection());
+        final Broadcast<RecalibrationReport> reportBroadcast = ctx.broadcast(bqsrReport);
+        final JavaRDD<GATKRead> finalReads = ApplyBQSRSparkFn.apply(markedReads, reportBroadcast, readsHeader, new ApplyBQSRArgumentCollection());
 
         try {
             ReadsSparkSink.writeReads(ctx, output, finalReads, readsHeader, ReadsWriteFormat.SHARDED);
```

Comment:
Added an option here as well to select between `ReadsWriteFormat.SINGLE` and `ReadsWriteFormat.SHARDED`